contrib/opencensus-ext-azure/opencensus/ext/azure/log_exporter/__init__.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2019, OpenCensus Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import threading
+import time
+
+from opencensus.common.schedule import Queue
+from opencensus.common.schedule import QueueEvent
+from opencensus.ext.azure.common import Options
+
+logger = logging.getLogger(__name__)
+
+__all__ = ['AzureLogHandler']
+
+
+class BaseLogHandler(logging.Handler):
+    def __init__(self):
+        super(BaseLogHandler, self).__init__()
+        self._queue = Queue(capacity=8192)  # TODO: make this configurable
+        self._worker = Worker(self._queue, self)
+        self._worker.start()
+
+    def close(self):
+        self._worker.stop()
+
+    def createLock(self):
+        self.lock = None
+
+    def emit(self, record):
+        self._queue.put(record, block=False)
+
+    def _export(self, batch, event=None):
+        try:
+            return self.export(batch)
+        finally:
+            if event:
+                event.set()
+
+    def export(self, batch):
+        raise NotImplementedError
+
+    def flush(self, timeout=None):
+        self._queue.flush(timeout=timeout)
+
+
+class Worker(threading.Thread):
+    daemon = True
+
+    def __init__(self, src, dst):
+        self._src = src
+        self._dst = dst
```

Comment:
Note the circular reference here, we've been bitten by this before trying to clean up exporters. See https://github.com/census-instrumentation/opencensus-python/blob/04c111f99dc122a079b589942b61d22c23979afa/contrib/opencensus-ext-stackdriver/opencensus/ext/stackdriver/stats_exporter/__init__.py#L372-L406.

Reply:
My original understanding is that we have `Handler.close` which can be used explicitly or implicitly (by Python logging library), and the circular reference will be collected by the GC. This shouldn't cause problem? Is this for preventing memory leak, or to reduce GC overhead?

Regarding memory leak, I run the following app for an hour and see a flat memory usage:
```
import logging

from opencensus.ext.azure.log_exporter import AzureLogHandler

logger = logging.getLogger(__name__)
while True:
    handler = AzureLogHandler()
    logger.addHandler(handler)
    logger.warning('Hello, World!')
    logger.removeHandler(handler)
    handler.close()
```

For GC overhead, given the circle is pretty small, I guess there is no noticeable difference?

Reply:
> the circular reference will be collected by the GC. This shouldn't cause problem?

I thought there were more general problems cleaning up circular references in python, but it may only be a problem for classes that define `__del__` (see https://stackoverflow.com/a/2428888), in which case this is fine.